pkg/network/server.go

```diff
@@ -563,27 +580,63 @@ func (s *Server) getVersionMsg() (*Message, error) {
 func (s *Server) IsInSync() bool {
 	var peersNumber int
 	var notHigher int
+	/*
+		if !s.stateSync.IsActive() {
+			s.bSyncQueue.discard()
+		}
+	*/
 
-	if s.MinPeers == 0 {
+	if s.MinPeers == 0 && !s.stateSync.IsActive() {
 		return true
 	}
 
 	ourLastBlock := s.chain.BlockHeight()
 
 	s.lock.RLock()
+	heights := make([]uint32, 0)
 	for p := range s.peers {
 		if p.Handshaked() {
 			peersNumber++
-			if ourLastBlock >= p.LastBlockIndex() {
+			peerLastBlock := p.LastBlockIndex()
+			if ourLastBlock >= peerLastBlock {
 				notHigher++
 			}
+			i := sort.Search(len(heights), func(i int) bool {
+				return heights[i] >= peerLastBlock
+			})
+			heights = append(heights, peerLastBlock)
+			if i != len(heights)-1 {
+				copy(heights[i+1:], heights[i:])
+				heights[i] = peerLastBlock
+			}
 		}
 	}
 	s.lock.RUnlock()
 
+	if s.stateSync.IsActive() && peersNumber >= s.MinPeers && len(heights) > 0 {
+		// choose the height of the median peer as current chain's height
+		h := heights[len(heights)/2]
+		err := s.stateSync.Init(h, func(m map[util.Uint256]bool) error {
```

Comment:
We're changing state in `IsInSync` again, we shouldn't be doing that.

Reply:
But `IsInSync` is the first place where I can get around the needed set of peers to initialise the StateSync module.

Reply:
I'll try to move this logic to another place.

Reply:
Moved to a separate method (`(* Server).tryInitStateSync`), this method is called each time `CMDVerack` is successfully handled.